aot/c_codegen.py

```diff
@@ -0,0 +1,213 @@
+        return sig
+
+
+def py_str_to_uchar_array(txt: str) -> Tuple[str, int]:  # (c_code, array len)
+    """Hexdump as string into a C array"""
+    hex_ = str(binascii.hexlify(bytes(txt, "utf-8")))[2:-1]
+    it = iter(hex_)
+    data = ", ".join([f"0x{x}{next(it)}" for x in it])
+    return data, len(hex_)
+
+
+# CUdeviceptr x, CUdeviceptr y, CUdeviceptr res, size_t size
+def gen_kernel_code(name: str, ptx_src: str, data_sign: CInputSigniture) -> str:
+    ptx_as_hex, ptx_size = py_str_to_uchar_array(ptx_src)
+
+    ptx_def = f"unsigned char {name}_ptx[{ptx_size}]"
+    kernel_fn_def = (
+        f"CUresult {name}(CUstream stream, GridWarps g, {data_sign.signitue_c_code()})"
+    )
+
+    get_func_def = f"CUfunction {name}_fn()"
+
+    # === Kernel header ===
+    header_src = [f"{ptx_def};", f"{get_func_def};", f"{kernel_fn_def};"]
+
+    # === PTX data ===
+    ptx_data_src = [f"{ptx_def} = ", "{", f" {ptx_as_hex}", "};"]
+
+    # === Functions ===
+    _if_error = ["if (err != 0)", "{", "\treturn NULL;", "}"]
+
+    def _func_wrapper(fn_def, fn_body):
+        return [f"{fn_def}", "{", *fn_body, "}"]
+
+    # ==== Load Kernel Func ====
+    load_kernel_body = [
+        "CUmodule mod_ptr;",
+        "CUfunction func;",
+        "CUresult err;",
+        f"void *image = (void *)&{name}_ptx;",
+        "err = cuModuleLoadData(&mod_ptr, image);",
+        *_if_error,
+        f'err = cuModuleGetFunction(&func, mod_ptr, "{name}");',
+        *_if_error,
+        "return func;",
+    ]
+
+    load_kernel_fn_src = _func_wrapper(get_func_def, load_kernel_body)
+
+    # ==== Run Kernel Func ====
+    args = "\n".join(["{", ", ".join([f"&{arg}" for arg in data_sign.arg_names]), "}"])
+
+    kernel_fn_body = [
+        f"CUfunction func = {name}_fn();",
+        f"void *args[{len(data_sign)}] = {args};",
+        "return cuLaunchKernel(func, g.gX, g.gY, g.gZ, g.numWarps * 32, 1, 1, 0, stream, args, NULL);",
+    ]
+
+    kernel_fn_src = _func_wrapper(kernel_fn_def, kernel_fn_body)
+
+    return header_src, ptx_data_src, load_kernel_fn_src, kernel_fn_src
+
+
+class SharedLibrarySource:
+    def __init__(self, lib_name: str, output_dir=None) -> None:
+
+        self._lib_name = lib_name
+        self._output_dir = output_dir
+
+        self.header_tmplt = [
+            "#include <stdio.h>",
+            "#include <stdint.h>",
+            "#include <inttypes.h>",
+            "#include <cuda.h>",
+            "", "",
+            "typedef struct",
+            "{",
+            "int gX;",
+            "int gY;",
+            "int gZ;",
+            "int numWarps;",
+            "} GridWarps;",
+            "", "",
+        ]
+
+        self.src_ptx_data = []
+        self.src_get_funcs = []
+        self.src_launch_funcs = []
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self):
+        return False
+
+    def _wrapped_header(self):
+        return (
+            ["#ifndef TRITON_KERNELS_H", "#define TRITON_KERNEL_H"]
```

Comment:
I think we should use `#pragma once` here, since it is widely supported across all major compilers. Coming up with unique names for header guards will be tricky since users are also likely to move the header around.